Adopt smart pointers in WebWheelEventCoalescer, SpeechRecognitionRealtimeMediaSourceManager, WebServiceWorkerFetchTaskClient

jelee53 opened the pull request:
#### d3757cff5f3f43d667153f39de08f842bf90b5b0
<pre>
Adopt smart pointers in WebWheelEventCoalescer, SpeechRecognitionRealtimeMediaSourceManager, WebServiceWorkerFetchTaskClient
<a href="https://bugs.webkit.org/show_bug.cgi?id=296119">https://bugs.webkit.org/show_bug.cgi?id=296119</a>
<a href="https://rdar.apple.com/156041248">rdar://156041248</a>

Reviewed by Basuke Suzuki.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>

* Source/WebCore/Modules/fetch/FormDataConsumer.cpp:
(WebCore::FormDataConsumer::consumeBlob):
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::~Blob):
* Source/WebCore/fileapi/BlobLoader.h:
* Source/WebCore/fileapi/NetworkSendQueue.cpp:
(WebCore::NetworkSendQueue::enqueue):
* Source/WebCore/page/ShareDataReader.cpp:
(WebCore::ShareDataReader::start):
(WebCore::ShareDataReader::didFinishLoading):
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Shared/WebWheelEventCoalescer.cpp:
(WebKit::WebWheelEventCoalescer::nextEventToDispatch):
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::~SpeechRecognitionRealtimeMediaSourceManager):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::start):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::stop):
* Source/WebKit/WebProcess/Storage/WebServiceWorkerFetchTaskClient.cpp:
(WebKit::WebServiceWorkerFetchTaskClient::didReceiveFormDataAndFinishInternal):

Canonical link: <a href="https://commits.webkit.org/298799@main">https://commits.webkit.org/298799@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4d6f2a511b8ef1ef1330083656906caa1c2ff286

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116680 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36344 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26906 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122754 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67252 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/60cb2589-ce18-497a-b8eb-413342795358) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118569 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37042 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44933 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88612 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43024 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/ba20327e-f241-421c-8020-5e2134c5b801) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29542 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104675 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69079 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28604 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22781 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66421 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98922 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22936 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125890 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43579 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/32732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97279 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43943 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100878 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97073 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42399 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20330 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/39541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18632 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43465 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49060 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42932 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46271 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44637 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jelee53 commented on Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
```diff
@@ -160,7 +162,7 @@ SpeechRecognitionRealtimeMediaSourceManager::SpeechRecognitionRealtimeMediaSourc
 SpeechRecognitionRealtimeMediaSourceManager::~SpeechRecognitionRealtimeMediaSourceManager()
 {
-    m_process->removeMessageReceiver(*this);
+    CheckedRef { m_process.get() }->removeMessageReceiver(*this);
```
I'm going to leave the original member variable `WebProcess m_process` as a WeakRef for now. I think it's better to leave the member variable as a WeakRef just because I don't think that a SpeechRecognitionManager should prevent WebProcess from being destroyed (memory zero'd out). 

If it was a CheckedRef and SpeechRecognitionManager was the last one to observe this particular WebProcess, the checkedCount may prevent the WebProcess memory from being cleared.

basuke commented on Source/WebCore/Modules/fetch/FormDataConsumer.cpp:
```diff
@@ -113,10 +113,10 @@ void FormDataConsumer::consumeBlob(const URL& blobURL)
             protectedThis->consume(data->span());
     });
+    CheckedPtr blobLoader = m_blobLoader.get();
+    blobLoader->start(blobURL, m_context.get(), FileReaderLoader::ReadAsArrayBuffer);
```
You didn't check blobLoader is not null before use.

RupinMittal replied:
I don't think a null check is needed. `blobLoader` is guaranteed to exist at this point because it's created right before this with `makeUnique`. This is likely why the previous code didn't have a null check before calling `start` either.

Although something about this seems off. The fact that this code makes the object and then null checks it below makes it seem like they expect that calling `start` may destroy the object. Is it ok to have a `CheckedPtr` to it in that case? @rniwa

RupinMittal commented on Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
```diff
@@ -133,10 +134,11 @@ class SpeechRecognitionRealtimeMediaSourceManager::Source final
-        if (AudioSession::singleton().preferredBufferSize() > bufferSize)
-            AudioSession::singleton().setPreferredBufferSize(bufferSize);
-        AudioSession::singleton().setCategory(AudioSession::CategoryType::PlayAndRecord, AudioSession::Mode::Default, RouteSharingPolicy::Default);
+        Ref audioSesion = AudioSession::singleton();
```
If it's a singleton we shouldn't need to ref it. Perhaps we can just do `auto audioSessionSingleton = AudioSession::singleton();` and use that below. Although it's kind of odd that AudioSession is a singleton but also has a public `create` function that returns non-singleton AudioSessions. Is that ok? @rniwa

basuke commented on Source/WebCore/Modules/fetch/FormDataConsumer.cpp:
```diff
@@ -113,10 +113,12 @@ void FormDataConsumer::consumeBlob(const URL& blobURL)
+        if (!blobLoader->isLoading())
+            didFail(Exception { ExceptionCode::InvalidStateError, "Unable to read form data blob"_s });
+    } else
         didFail(Exception { ExceptionCode::InvalidStateError, "Unable to read form data blob"_s });
```
Can you do something like:
```
if (CheckedPtr blobLoader = m_blobLoader.get()) {
  blobLoader->start(...);
  if (blobLoader->isLoading())
    return;
}
didFail(...);
```

basuke replied:
Even if you don't do above, double parentheses are not good.

jelee53 replied:
That code is a lot prettier than mine. I should have cleaned up more carefully before putting up the PR. 😅

(I was getting build errors from the extra parentheses but they went away. Not sure what the issue there was!)

jelee53 replied:
Done!

jelee53 committed:
Adopt smart pointers in WebWheelEventCoalescer, SpeechRecognitionRealtimeMediaSourceManager, WebServiceWorkerFetchTaskClient
https://bugs.webkit.org/show_bug.cgi?id=296119
rdar://156041248

Reviewed by Basuke Suzuki.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines

* Source/WebCore/Modules/fetch/FormDataConsumer.cpp:
(WebCore::FormDataConsumer::consumeBlob):
* Source/WebCore/fileapi/Blob.cpp:
(WebCore::Blob::~Blob):
* Source/WebCore/fileapi/BlobLoader.h:
* Source/WebCore/fileapi/NetworkSendQueue.cpp:
(WebCore::NetworkSendQueue::enqueue):
* Source/WebCore/page/ShareDataReader.cpp:
(WebCore::ShareDataReader::start):
(WebCore::ShareDataReader::didFinishLoading):
* Source/WebKit/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncheckedLocalVarsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Shared/WebWheelEventCoalescer.cpp:
(WebKit::WebWheelEventCoalescer::nextEventToDispatch):
* Source/WebKit/WebProcess/Speech/SpeechRecognitionRealtimeMediaSourceManager.cpp:
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::~SpeechRecognitionRealtimeMediaSourceManager):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::start):
(WebKit::SpeechRecognitionRealtimeMediaSourceManager::stop):
* Source/WebKit/WebProcess/Storage/WebServiceWorkerFetchTaskClient.cpp:
(WebKit::WebServiceWorkerFetchTaskClient::didReceiveFormDataAndFinishInternal):

Canonical link: https://commits.webkit.org/298799@main
